Implement ID3DUserDefinedAnnotations for D3D11

liam-middlebrook opened the pull request:
I plan to follow-up with the D3D9-era `D3DPERF_*` implementation in a separate PR. That implementation isn't quite a straight-forward given that those functions are global and operate outside of any device/instance context.

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
           LPCWSTR                 Name) {
-    // Currently not implemented
-    return -1;
+    D3D11DeviceContext *context = dynamic_cast<D3D11DeviceContext *>(m_container);
```
Instead of this, just make `m_container` a `D3D11DeviceContext` (forward declare the class in the header).

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
+
+    if (context) {
+      context->EmitCs([labelName = dxvk::str::fromws(Name)](DxvkContext *ctx) {
+        VkDebugUtilsLabelEXT label = { .sType = VK_STRUCTURE_TYPE_DEBUG_UTILS_LABEL_EXT };
```
This is a C++20 feature :(

Sorry

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
+        VkDebugUtilsLabelEXT label = { .sType = VK_STRUCTURE_TYPE_DEBUG_UTILS_LABEL_EXT };
+        label.pLabelName = labelName.c_str();
+
+        ctx->beginDebugLabel(
```
ctx->beginDebugLabel(&label);

Joshua-Ashton commented on src/dxvk/dxvk_extensions.h:
```diff
@@ -292,8 +292,9 @@ namespace dxvk {
    * used by DXVK if supported by the implementation.
    */
   struct DxvkInstanceExtensions {
+    DxvkExt extDebugUtils                   = { VK_EXT_DEBUG_UTILS_EXTENSION_NAME,                      DxvkExtMode::Optional };
```
We shouldn't enable this by default. It's not free.

Joshua-Ashton replied:
Maybe use an env var for this like we do in VKD3D.

DadSchoorse replied:
Or a dxvk.conf option.

liam-middlebrook replied:
I'd prefer to go with an environment variable for toggling this on/off, since it's a bit more lightweight to enable while changing between different applications. How does a boolean `DXVK_PERF_EVENTS` sound?

Joshua-Ashton replied:
Sounds good.

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
-    return -1;
+    D3D11DeviceContext *context = dynamic_cast<D3D11DeviceContext *>(m_container);
+
+    if (context) {
```
We always have a context. We're a QueryInterface of it!

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
-    return -1;
+    D3D11DeviceContext *context = dynamic_cast<D3D11DeviceContext *>(m_container);
+
+    if (context) {
```
We always have a context. We're a QueryInterface of it!

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
-    // Currently not implemented
+    D3D11DeviceContext *context = dynamic_cast<D3D11DeviceContext *>(m_container);
+
+    if (context) {
```
We always have a context. We're a QueryInterface of it!

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
+        VkDebugUtilsLabelEXT label = { .sType = VK_STRUCTURE_TYPE_DEBUG_UTILS_LABEL_EXT };
+        label.pLabelName = labelName.c_str();
+
+        ctx->insertDebugLabel(
```
ctx->insertDebugLabel(&label);

Joshua-Ashton commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,59 @@ namespace dxvk {
-    // Currently not implemented
+    D3D11DeviceContext *context = dynamic_cast<D3D11DeviceContext *>(m_container);
+
+    if (context) {
```
You get the picture.

Joshua-Ashton reviewed (changes requested):
Another thing: I'd prefer having a !extDebugUtils check on the frontend also, that way, if games end up using this, we don't have a bunch of overhead from allocating a string and copying into that and pushing that on the CS thread for it to never be used.

liam-middlebrook commented:
> Another thing: I'd prefer having a !extDebugUtils check on the frontend also, that way, if games end up using this, we don't have a bunch of overhead from allocating a string and copying into that and pushing that on the CS thread for it to never be used.

Sure I think that makes sense. Especially since applications _can_ still end these perf events even if the mechanism isn't reported as enabled/supported.

doitsujin commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -1,9 +1,12 @@
 
   D3D11UserDefinedAnnotation::D3D11UserDefinedAnnotation(ID3D11DeviceContext* ctx)
-  : m_container(ctx) { }
+  : m_container(ctx),
```
Now that we actually interact with the implementation, `ctx` should be changed to `D3D11DeviceContext` to avoid the useless casts.

liam-middlebrook replied:
Sure.

doitsujin commented on src/d3d11/d3d11_annotation.h:
```diff
@@ -33,6 +33,9 @@ namespace dxvk {
     ID3D11DeviceContext*  m_container;
 
+    // Stack depth for non-finalized BeginEvent calls
+    int m_eventDepth;
```
is there a reason why this is signed? (also, prefer `int32_t` or `uint32_t`)

Joshua-Ashton replied:
I guess because D3D returns this signed

liam-middlebrook replied:
Sure. I'll make this a sized type. Josh is correct, it's signed because that's how the D3D interface wants it. Basically it's a stack-like behavior of events being pushed/popped and if we have a negative number that indicates an error from the application ending more event regions than it began.

doitsujin commented on src/dxvk/dxvk_cmdlist.cpp:
```diff
@@ -206,4 +206,15 @@ namespace dxvk {
-}
\ No newline at end of file
+  void DxvkCommandList::cmdBeginDebugUtilsLabel(VkDebugUtilsLabelEXT *pLabelInfo) {
+    m_device->instance()->vki()->vkCmdBeginDebugUtilsLabelEXT(m_execBuffer, pLabelInfo);
```
we could also store the instance dispatch table in the command list somewhere to make this a bit less silly

liam-middlebrook replied:
Sure.

doitsujin reviewed (changes requested):
See the things Josh pointed out as well.

liam-middlebrook commented:
Okay, I've addressed the feedback from above and added a new environment variable `DXVK_PERF_EVENTS` for enabling this functionality for use with debugging/profiling tools.

doitsujin commented on src/dxvk/dxvk_instance.cpp:
```diff
@@ -105,6 +105,26 @@ namespace dxvk {
+      extensionsEnabled.merge(utilsExtensionsEnabled);
+    }
+
+    m_extensions = insExtensions;
```
I think a much easier way to do this would be to turn `insExtensionList` into an `std::vector` and conditionally push the pointer to it. The whole extension merging thing is meant for extensions that we don't necessarily know about.

liam-middlebrook replied:
Okay that makes sense. So the merging is for the "plugins" like for OpenVR/OpenXR or DXVK Extensions like dxvk-ags or dxvk-nvapi ? Either way, fixed.

doitsujin commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,60 @@ namespace dxvk {
+    m_container->EmitCs([labelName = dxvk::str::fromws(Name)](DxvkContext *ctx) {
+      VkDebugUtilsLabelEXT label;
+      label.sType = VK_STRUCTURE_TYPE_DEBUG_UTILS_LABEL_EXT;
+      label.pNext = NULL;
```
`nullptr`

doitsujin commented on src/d3d11/d3d11_annotation.cpp:
```diff
@@ -30,26 +33,60 @@ namespace dxvk {
+    m_container->EmitCs([labelName = dxvk::str::fromws(Name)](DxvkContext *ctx) {
+      VkDebugUtilsLabelEXT label;
+      label.sType = VK_STRUCTURE_TYPE_DEBUG_UTILS_LABEL_EXT;
+      label.pNext = NULL;
```
`nullptr`

doitsujin reviewed (changes requested):
Small nits, looks good otherwise.

Joshua-Ashton reviewed (approved):
LGTM
